### PR TITLE
fix: block `mload` merging when src and dst overlap

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fix an optimization bug introduced in #3483 when the target architecture is `cancun`. prior to this PR, certain sequences of `mload/mstore` could incorrectly be merged into `mcopy` when there is overlap between the source and destination buffers.

### Commit message

```
this commit fixes an optimization bug when the target architecture has
the `mcopy` instruction (i.e. `cancun` or later). the bug was introduced
in 5dc3ac7. specifically, the `merge_mload` step can incorrectly merge
`mload`/`mstore` sequences (into `mcopy`) when the source and
destination buffers overlap, and the destination buffer is "ahead of"
(i.e. greater than) the source buffer. this commit fixes the issue by
blocking the optimization in these cases, and adds unit and functional
tests demonstrating the correct behavior.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ir/optimizer.py [hunk 1]
```diff
@@ -662,10 +662,10 @@ def _rewrite_mstore_dload(argz):
 def _merge_mload(argz):
     if not version_check(begin="cancun"):
         return False
-    return _merge_load(argz, "mload", "mcopy")
+    return _merge_load(argz, "mload", "mcopy", allow_overlap=False)
 
 
-def _merge_load(argz, _LOAD, _COPY):
+def _merge_load(argz, _LOAD, _COPY, allow_overlap=True):
```
Comment (chen-robert) [inline vyper/ir/optimizer.py:668]: nit: think generally params should be "safe-by-default", and the "safer" option would be to have no overlap.

File: vyper/ir/optimizer.py [hunk 2]
```diff
@@ -689,6 +689,11 @@ def _merge_load(argz, _LOAD, _COPY):
                 initial_dst_offset = dst_offset
                 initial_src_offset = src_offset
                 idx = i
+
+            if not allow_overlap and initial_dst_offset <= initial_src_offset <= dst_offset:
```
Comment (chen-robert) [inline vyper/ir/optimizer.py:693]: do you have to do do a + 32 somewhere here?

Comment (charles-cooper) [reply vyper/ir/optimizer.py]: i think the `<=` handles that, see the output of this "partial" overlap:

```vyper
@external
def partial() -> uint256:
    a: uint256 = 1337
    b: uint256 = 0
    c: uint256 = 0
    d: uint256 = 0
    e: uint256 = 0

    c = a
    d = b
    e = c

    return e
```

File: vyper/ir/optimizer.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/ir/optimizer.py:668]: hmm i see, my thinking was that this only affects where the source and destination address spaces are the same, and there is only one case for this, `mload` / `mstore`. but i am open to changing it.

**Review by chen-robert — COMMENTED** [review]
lgtm

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**